azurerm/data_source_azuread_service_principal.go

```diff
@@ -55,47 +57,61 @@ func dataSourceArmActiveDirectoryServicePrincipalRead(d *schema.ResourceData, me
 		}
 
 		servicePrincipal = &app
-	} else {
-		apps, err := client.ListComplete(ctx, "")
+
+	} else if _, ok := d.GetOk("display_name"); ok {
+
+		// use the display_name to find the Azure AD service principal
+		displayName := d.Get("display_name").(string)
+		filter := "displayName eq '" + displayName + "'"
+		log.Printf("[DEBUG] [data_source_azuread_service_principal] Using filter %q", filter)
+
+		apps, err := client.ListComplete(ctx, filter)
 		if err != nil {
 			return fmt.Errorf("Error listing Service Principals: %+v", err)
 		}
 
-		if v, ok := d.GetOk("display_name"); ok {
-			displayName := v.(string)
-
-			for _, app := range *apps.Response().Value {
-				if app.DisplayName == nil {
-					continue
-				}
-
-				if *app.DisplayName == displayName {
-					servicePrincipal = &app
-					break
-				}
+		for _, app := range *apps.Response().Value {
+			if app.DisplayName == nil {
+				continue
 			}
 
-			if servicePrincipal == nil {
-				return fmt.Errorf("A Service Principal with the Display Name %q was not found", displayName)
+			if *app.DisplayName == displayName {
+				servicePrincipal = &app
+				break
 			}
-		} else {
-			applicationId := d.Get("application_id").(string)
-
-			for _, app := range *apps.Response().Value {
-				if app.AppID == nil {
-					continue
-				}
-
-				if *app.AppID == applicationId {
-					servicePrincipal = &app
-					break
-				}
+		}
+
+		if servicePrincipal == nil {
+			return fmt.Errorf("A Service Principal with the Display Name %q was not found", displayName)
+		}
+
+	} else {
+
+		// use the application_id to find the Azure AD service principal
+		applicationId := d.Get("application_id").(string)
+		filter := "appId eq '" + applicationId + "'"
```

Comment:
(as above) could we use `fmt.Sprintf("appId eq '%q'", applicationId)` here?

Reply:
Makes sense! Added as requested.